File: src/lib/repositories/tickets.ts
import type { Ticket, TicketStatus } from '../types';
import type { TicketRepository } from './interfaces';

type TicketRow = {
  id: number;
  event_id: number;
  seat_number: string;
  section: string | null;
  row_number: string | null;
  status: TicketStatus;
  version: number | null;
  created_at: Date;
  updated_at: Date;
};

const COLUMNS = 'id, event_id, seat_number, section, row_number, status, version, created_at, updated_at';

function toTicket(row: TicketRow): Ticket {
  return {
    id: row.id,
    eventId: row.event_id,
    seatNumber: row.seat_number,
    section: row.section,
    rowNumber: row.row_number,
    status: row.status,
    version: row.version ?? 0,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export const ticketRepository: TicketRepository = {
  async findByIdForUpdate(q, id) {
    const res = await q.query<TicketRow>(`SELECT ${COLUMNS} FROM tickets WHERE id = $1 FOR UPDATE`, [id]);
    const row = res.rows[0];
    return row ? toTicket(row) : null;
  },

  async findByEventAndSeatForUpdate(q, eventId, seatNumber) {
    const res = await q.query<TicketRow>(
      `SELECT ${COLUMNS} FROM tickets WHERE event_id = $1 AND seat_number = $2 FOR UPDATE`,
      [eventId, seatNumber],
    );
    const row = res.rows[0];
    return row ? toTicket(row) : null;
  },

  async listByEvent(q, eventId) {
    const res = await q.query<TicketRow>(`SELECT ${COLUMNS} FROM tickets WHERE event_id = $1 ORDER BY id`, [eventId]);
    return res.rows.map(toTicket);
  },

  async updateStatus(q, id, status) {
    await q.query('UPDATE tickets SET status = $2, version = COALESCE(version, 0) + 1 WHERE id = $1', [id, status]);
  },
};

File: backend/app/repositories/tickets.py
"""Equivalente a repositories/tickets.ts."""
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Ticket


async def find_by_id_for_update(session: AsyncSession, ticket_id: int) -> Ticket | None:
    result = await session.execute(select(Ticket).where(Ticket.id == ticket_id).with_for_update())
    return result.scalar_one_or_none()


async def find_by_event_and_seat_for_update(session: AsyncSession, event_id: int, seat_number: str) -> Ticket | None:
    result = await session.execute(
        select(Ticket).where(Ticket.event_id == event_id, Ticket.seat_number == seat_number).with_for_update()
    )
    return result.scalar_one_or_none()


async def list_by_event(session: AsyncSession, event_id: int) -> list[Ticket]:
    result = await session.execute(select(Ticket).where(Ticket.event_id == event_id).order_by(Ticket.id))
    return list(result.scalars().all())


async def update_status(session: AsyncSession, ticket_id: int, status: str) -> None:
    ticket = await session.get(Ticket, ticket_id)
    if ticket is not None:
        ticket.status = status
        ticket.version = (ticket.version or 0) + 1
